litex_boards/targets/zcu104.py

```diff
@@ -0,0 +1,107 @@
+#!/usr/bin/env python3
+import argparse
+
+from migen import *
+from migen.genlib.io import CRG
+
+from litex_boards.platforms import zcu104
+
+from litex.soc.cores.clock import *
+from litex.soc.integration.soc_sdram import *
+from litex.soc.integration.builder import *
+
+from litedram.modules import EDY4016A
+from litedram.phy import usddrphy
+
+# BaseSoC ------------------------------------------------------------------------------------------
+
+class _CRG(Module):
+    def __init__(self, platform, sys_clk_freq):
+        self.clock_domains.cd_sys    = ClockDomain()
+        self.clock_domains.cd_sys4x  = ClockDomain(reset_less=True)
+        self.clock_domains.cd_pll4x  = ClockDomain(reset_less=True)
+        self.clock_domains.cd_clk200 = ClockDomain()
+        self.clock_domains.cd_ic     = ClockDomain()
+
+        self.submodules.pll = pll = USMMCM(speedgrade=-2)
+        pll.register_clkin(platform.request("clk125"), 125e6)
+
+        pll.create_clkout(self.cd_pll4x, sys_clk_freq*4, buf=None, with_reset=False)
+        pll.create_clkout(self.cd_clk200, 200e6, with_reset=False)
+
+        self.specials += [
+            Instance("BUFGCE_DIV", name="main_bufgce_div",
+                p_BUFGCE_DIVIDE=4,
+                i_CE=1, i_I=self.cd_pll4x.clk, o_O=self.cd_sys.clk),
+            Instance("BUFGCE", name="main_bufgce",
+                i_CE=1, i_I=self.cd_pll4x.clk, o_O=self.cd_sys4x.clk),
+            AsyncResetSynchronizer(self.cd_clk200, ~pll.locked),
+        ]
+
+        ic_reset_counter = Signal(max=64, reset=63)
+        ic_reset = Signal(reset=1)
+        self.sync.clk200 += \
+            If(ic_reset_counter != 0,
+                ic_reset_counter.eq(ic_reset_counter - 1)
+            ).Else(
+                ic_reset.eq(0)
+            )
+        ic_rdy = Signal()
+        ic_rdy_counter = Signal(max=64, reset=63)
+        self.cd_sys.rst.reset = 1
+        self.comb += self.cd_ic.clk.eq(self.cd_sys.clk)
+        self.sync.ic += [
+            If(ic_rdy,
+                If(ic_rdy_counter != 0,
+                    ic_rdy_counter.eq(ic_rdy_counter - 1)
+                ).Else(
+                    self.cd_sys.rst.eq(0)
+                )
+            )
+        ]
+        self.specials += [
+            Instance("IDELAYCTRL", p_SIM_DEVICE="ULTRASCALE",
+                     i_REFCLK=ClockSignal("clk200"), i_RST=ic_reset,
```

Comment:
Although I didn't realise at the time and 200MHz worked fine; officially the UltraScale+ requires 300MHz minimum for IDELAYCTRLs, this may matter if pushing the DDR4 clock higher and needing finer delay control.